#172 Resource detectors for Docker

swetharavichandrancisco opened the pull request:
- Added Resource detectors for Docker environment to extract container.id
- Added Unit Tests

cijothomas commented:
@swetharavichandrancisco could you resolve conflicts and rebase?

cijothomas commented on src/OpenTelemetry.Contrib.Extensions.Docker/OpenTelemetry.Contrib.Extensions.Docker.csproj:
```diff
@@ -0,0 +1,9 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>;net461;netstandard2.0</TargetFrameworks>
+    <Description>OpenTelemetry extensions for Docker</Description>
```
bit more descriptive, as this is very specifically for Resource detection.

utpilla commented:
Please remove the word `Contrib` from the file names, folder names, namespaces, etc. We have removed the word `Contrib` for rest of the packages in the repo: #196

utpilla commented on src/OpenTelemetry.Contrib.Extensions.Docker/Resources/DockerResourceDetector.cs:
```diff
@@ -0,0 +1,127 @@
+    /// </summary>
+    public class DockerResourceDetector : IResourceDetector
+    {
+        private const string FILEPATH = "/proc/self/cgroup";
```
Which is better to use here? `readonly' or `const`?

utpilla commented on test/OpenTelemetry.Contrib.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        [Fact]
+        public void TestGetIdFromLine()
```
Could we simply test for the different possible test cases using `Resource.Attributes` instead of testing the helper methods?

I think we should mark the methods: `ExtractContainerId` and `GetIdFromLine` as `private`.

utpilla commented on test/OpenTelemetry.Contrib.Extensions.Docker.Tests/Resources/SampleFile/testcgroup:
```diff
@@ -0,0 +1,14 @@
+14:name=systemd:/docker/da261b882cd310e64ced2e31b938f9289261e4c7d59efb2d651ca2fe802b7764
```
What's the purpose of this file?

utpilla commented:
Add CHANGELOG.md and update the `component_owners.yml` file for this project.

utpilla commented on src/OpenTelemetry.Contrib.Extensions.Docker/DockerEventSource.cs:
```diff
@@ -0,0 +1,63 @@
+
+namespace OpenTelemetry.Contrib.Extensions.Docker
+{
+    [EventSource(Name = "OpenTelemetry-Docker")]
```
Use `OpenTelemetry-Extensions-Docker` as the EventSource name.

utpilla commented on src/OpenTelemetry.Contrib.Extensions.Docker/DockerEventSource.cs:
```diff
@@ -0,0 +1,63 @@
+﻿// <copyright file="DockerEventSource.cs" company="OpenTelemetry Authors">
```
nit: Rename the file to `DockerExtensionsEventSource.cs`

utpilla commented on src/OpenTelemetry.Contrib.Extensions.Docker/DockerEventSource.cs:
```diff
@@ -0,0 +1,63 @@
+        [NonEvent]
+        public void ResourceAttributesExtractException(string format, Exception ex)
+        {
+            if (this.IsEnabled(EventLevel.Warning, (EventKeywords)(-1)))
```
Use `EventLevel.Error`?

utpilla commented on test/OpenTelemetry.Contrib.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs:
```diff
@@ -0,0 +1,86 @@
+            Assert.Equal(CONTAINERID, dockerResourceDetector.GetIdFromLine(CGROUPLINE));
+        }
+    }
+}
```
Add a test case for the case when containerId is not found and it's returned as `null`.

utpilla commented on src/OpenTelemetry.Extensions.Docker/README.md:
```diff
@@ -0,0 +1,38 @@
+                        // other configurations
+                        .SetResourceBuilder(ResourceBuilder
+                            .CreateDefault()
+                            .AddDockerDetector())
```
Where is the extension method defined?

utpilla replied:
Name the method `AddDockerResourceDetector`.

utpilla commented on src/OpenTelemetry.Extensions.Docker/DockerExtensionsEventSource.cs:
```diff
@@ -0,0 +1,63 @@
+        public static DockerExtensionsEventSource Log = new DockerExtensionsEventSource();
+
+        [NonEvent]
+        public void ResourceAttributesExtractException(string format, Exception ex)
```
nit: Rename the method to `ExtractResourceAttributesException`.

utpilla commented on test/OpenTelemetry.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs:
```diff
@@ -0,0 +1,112 @@
+using OpenTelemetry.Resources;
+using Xunit;
+
+namespace OpenTelemetry.Extensions.Docker.Tests.Resources
```
The namespace should be `OpenTelemetry.Extensions.Docker.Tests`. The test explorer on Visual Studio would use the class name to categorize the different unit tests under the namespace.

utpilla commented on test/OpenTelemetry.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs:
```diff
@@ -0,0 +1,112 @@
+
+namespace OpenTelemetry.Extensions.Docker.Tests.Resources
+{
+    public class TestDockerResourceDetector
```
Rename this to `DockerResourceDetectorTests`. That's the convention followed in the contrib repo.

utpilla commented on src/OpenTelemetry.Extensions.Docker/OpenTelemetry.Extensions.Docker.csproj:
```diff
@@ -0,0 +1,9 @@
+    <Description>OpenTelemetry Extensions - Container Resource Detector from Docker environment.</Description>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="OpenTelemetry" Version="1.2.0-rc4" />
```
You could use the latest stable version of OTel SDK: `1.2.0`.

utpilla commented on test/OpenTelemetry.Extensions.Docker.Tests/OpenTelemetry.Extensions.Docker.Tests.csproj:
```diff
@@ -0,0 +1,23 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\OpenTelemetry.Extensions.Docker\OpenTelemetry.Extensions.Docker.csproj" />
```
Use `$(RepoRoot)` instead of `..\..\`.

utpilla commented:
@swetharavichandrancisco Thanks for updating the PR. There have been a few updates to the Contributing guidelines for this repo since you initially created this PR. Please add issue template, CI flow, etc. Here is the link to the doc for your reference: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md#contributing-a-new-project.

swetharavichandrancisco commented:
@utpilla, I have followed the contributing guidelines (from https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md#contributing-a-new-project) for my repo and updated the same as well.

cijothomas commented on src/OpenTelemetry.Extensions.Docker/DockerExtensionsEventSource.cs:
```diff
@@ -0,0 +1,63 @@
+            }
+        }
+
+        [Event(1, Message = "Failed to extract resource attributes in '{0}'.", Level = EventLevel.Error)]
```
only 1 parameter?

swetharavichandrancisco replied:
It Prints out the LOG Level with filename and an actual exception. Do you expect more?

cijothomas commented on src/OpenTelemetry.Extensions.Docker/OpenTelemetry.Extensions.Docker.csproj:
```diff
@@ -0,0 +1,10 @@
+    <MinVerTagPrefix>Extensions.Docker-</MinVerTagPrefix>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="OpenTelemetry" Version="1.2.0" />
+  </ItemGroup>
```
we can use 1.3.0 now

cijothomas commented on src/OpenTelemetry.Extensions.Docker/OpenTelemetry.Extensions.Docker.csproj:
```diff
@@ -0,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>;net461;netstandard2.0</TargetFrameworks>
```
I'd suggest net462. net461 is out of support already.

cijothomas commented on src/OpenTelemetry.Extensions.Docker/README.md:
```diff
@@ -0,0 +1,38 @@
+The resource detectors will record the following metadata based on where
+your application is running:
+
+- **DockerResourceDetector**: container id.
```
would be nice to list the name of the attribute used

cijothomas commented on src/OpenTelemetry.Extensions.Docker/Resources/DockerSemanticConventions.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    internal static class DockerSemanticConventions
+    {
+        public const string AttributeContainerID = "container.id";
```
please add a link to the relevant OTel spec/convention

swetharavichandrancisco replied:
I have added readme file which provides the usage/description as a part of previous PR - https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Extensions.Docker

I'm not sure if you mean by updating the OpenTelemtry docs - https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md. If I need to update it here, Do I need to provide an example as well?

cijothomas commented on test/OpenTelemetry.Extensions.Docker.Tests/OpenTelemetry.Extensions.Docker.Tests.csproj:
```diff
@@ -0,0 +1,23 @@
+
+  <PropertyGroup>
+    <Description>Unit test project for Docker Detector for OpenTelemetry</Description>
+    <TargetFrameworks>netcoreapp3.1</TargetFrameworks>
```
add net6.0 too

cijothomas commented on test/OpenTelemetry.Extensions.Docker.Tests/OpenTelemetry.Extensions.Docker.Tests.csproj:
```diff
@@ -0,0 +1,23 @@
+  <PropertyGroup>
+    <Description>Unit test project for Docker Detector for OpenTelemetry</Description>
+    <TargetFrameworks>netcoreapp3.1</TargetFrameworks>
+    <TargetFrameworks Condition="$(OS) == 'Windows_NT'">$(TargetFrameworks);net461</TargetFrameworks>
```
```suggestion
    <TargetFrameworks Condition="$(OS) == 'Windows_NT'">$(TargetFrameworks);net462</TargetFrameworks>
```

cijothomas reviewed (approved):
Left few comments. Would be nice to address before 1st release.

utpilla commented:
> Left few comments. Would be nice to address before 1st release.

@swetharavichandrancisco Please address these comments in a follow-up PR.

swetharavichandrancisco commented:
@cijothomas @utpilla , I have addressed the review comments of upgrading .NET Framework/OpenTelemetry version in the follow-up PR - https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/432
